### ISSUE TITLE
How to play in the background when the phone enters low power mode

### ISSUE THREAD

**Issue by Tomcater** [androidx/media#842]
Hello, I'm implementing a novel player. Now I have implemented a MediaLibraryService, and then implemented a DataProvider. The DataProvider provides playback metadata for the Player. Due to business requirements, only one MediaItem can be loaded into the Player at a time. After listening to the STATE_ENDED of the player, it starts to obtain the data of the next chapter through the DataProvider request interface, and then resets it to the player.
The problem now is that after starting playback, the user locks the screen and the phone enters low power mode. After the playback ends, STATE_ENDED will cause the foreground service of MediaLibraryService to be suspended. As a result, the DataProvider cannot connect to the Internet to request the loading of the next chapter's data. What should I do to solve the problem?

**Comment by marcbaechinger** [comment]
> the user locks the screen and the phone enters low power mode

Can you explain in a bit more detail what you mean with `low power mode`? Do you have some documentation about this and what restrictions are in place when in this mode?

> As a result, the DataProvider cannot connect to the Internet to request the loading of the next chapter's data.

How is this `cannot connect to the internet` expressed by the app? Is it throwing an exception? Can you post this exception here? I'm asking because I'd expect that you also load media data from the internet and I don't understand with available information why you can access the internet for media data but not for other data.

> After listening to the STATE_ENDED of the player, it starts to obtain the data of the next chapter
> through the DataProvider request interface

I understand that with `business requirements` you mean non-technical restrictions that you can not remove. Whoever is in charge and has the power to think about these requirements, should take into account that waiting for  `STATE_ENDED` and then starting to load next chapter data results in a poor user experience including buffering and a gap in the audio book listening experience that users probably do not like. So I think what I have heard more often are business requirements that ask for the opposite which is giving the player the chance to buffer media data ahead of transitioning to the next item. If not business people, this makes users happy.

**Comment by Tomcater** [comment]
I'm glad you responded to my question so quickly
First: For low power consumption mode, you can [refer to this](https://developer.android.com/training/monitoring-device-state/doze-standby)  or search for Doze. In low-power mode, only foreground services can access the network.
Second: After the first chapter data is played, the data of the second chapter needs to be obtained from the server. At this time, the interface reports an error "Connection timeout". I am sure that there is no problem with the interface access, because the player is in the END state at this time. MediaSessionService will stop the foreground service, so the server cannot be accessed from the background.
Third: Loading the next resource in advance can be played, but when the content of the next chapter needs to be paid, for a good user experience I need to first notify the user that payment is required, and then call the unlocking interface. After the unlocking is successful, load the next resource. chapters, so I can only load data one chapter at a time

**Comment by marcbaechinger** [comment]
Thanks for the details. 

> Loading the next resource in advance can be played, but when the content of the next chapter needs to be paid, for a good user experience I need to first notify the user that payment is required, and then call the unlocking interface.

I don't think this is going to work if this involves stopping or pausing the player. You must stay in the foreground anyway, or you can't resume without user interaction from a foreground activvty (or notification) Once your service is not in the foreground anymore, you can not start playback again from the background. Once the service is taken off the foreground, the user needs to start it from a foreground activity or a notification or by any other means that are exempted from not being allowed to start a foreground service from the background.

I think I'd try to do this the other way around. When I load an audio book, then I'd check for the users license and put all the chapters into the player that the user is allowed to play. These chapters play seamslessly, without ever falling into `STATE_ENDED`. When the player falls into state ended, you can check whether playback is actually at the end or whether the user should pay. When playback actually ended you do nothing, the user can resume from the notification. When you need a user to subscribe or upgrade, then you need to guide your user to an subscription activity anyway (for instance with a notitication).

**Comment by Tomcater** [comment]
Yes, like you said we can load the free chapters into the player in advance, then it can play automatically in sequence. When reaching the paid chapter, the user is prompted to pay next time. But I want to realize that after prompting the user, if the user has sufficient balance, I need to automatically help the user unlock the next chapter and then continue playing. But after checking the source code, I found that the foreground service will stop as long as the player enters the END or PAUSE state. 
`boolean shouldRunInForeground(
      MediaSession session, boolean startInForegroundWhenPaused) {
    @Nullable MediaController controller = getConnectedControllerForSession(session);
    return controller != null
        && (controller.getPlayWhenReady() || startInForegroundWhenPaused)
        && (controller.getPlaybackState() == Player.STATE_READY
            || controller.getPlaybackState() == Player.STATE_BUFFERING);
  }`
Can I override the onUpdateNotification method and always pass true as the second parameter so that the foreground service will not stop no matter when?
`override fun onUpdateNotification(session: MediaSession, startInForegroundRequired: Boolean) {
        super.onUpdateNotification(session, true)
    }`

**Comment by marcbaechinger** [comment]
That's technically possible. You are not following best practices with this though. It's up to you to take that decision and to see whether this continues to work well now and in the future.

Your service will, for instance, never be terminated by the system that way even when not playing so your notification will stay as long as your app stops it. That's the main point of best practice that you violate. That works with the current state of Android but you would need to test this yourself with every OS update I think. If users are annoyed by your notification having the notification never removed without them starting the app again, they may complain. I can't know if that happens, but you likely behave differently. than others.

From the point of view of the system. The situation you are in if you do so is to have a foreground service that is attached to a session. The system knows you are in the foreground while not playing. The code as it is now is what we recommend to do to play well with the restrictions of starting foreground services up top API 34. We will make sure that the default implementation will work in the future as well. So basically, if you want to do that differently I can't stop you do this. You will be in charge of making sure this works if you deviate from what the library does and the system asks for. 


Personally I'd first make sure my requirements can't be met without manipulating the foreground state of the service. What comes to my mind and you may want to consider is monitoring `onMediaItemTransitioned()` and set `ExoPlayer.setPauseAtEndOfMediaItems(true/false)` or use [`a PlayerMessage`](https://developer.android.com/reference/androidx/media3/exoplayer/PlayerMessage) when at the last free chapter. Use a `ForwardingPlayer` that is aware of paid/free content to intercept user commands. I think with these tools you can always play except when the users has no access in which case it is fine to stop the player  and go off the foreground.
